### PR TITLE
Use the same method to generate DllImportAttribute in local methods and P/Invoke forwarders

### PR DESCRIPTION
**Pull request by jtschuster**
We should forward the LibraryImpot.StringMarshalling as DllImport.CharSet to local P/Invokes too, not just forwarded DllImport's.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs
Comment (jkoritzinsky) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:449]: Instead of passing optional arguments, could we make a new instance of `libraryImportData` using with-expressions to set the right settings? And then also comment here about why we're setting particular settings (setting the entry point name and removing SetLastError).

**Comment by jkoritzinsky** [comment]
This is only applicable on the dotnet/runtime-only downlevel support, where we'll generate a source-generated stub for whatever parameters we can and forward whatever parameters we can't marshal (technically we also do this in main, but we emit errors in that case).

**Comment by jtschuster** [comment]
> This is only applicable on the dotnet/runtime-only downlevel support, where we'll generate a source-generated stub for whatever parameters we can and forward whatever parameters we can't marshal (technically we also do this in main, but we emit errors in that case).

Is that true? It looks like we have tests that expect a forwarder stub if we can't marshal all the parameters: https://github.com/dotnet/runtime/blob/c5e8f83db1606a89e70a9f4a0ac2fc03dea2418c/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/Compiles.cs#L524-L531

File: src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/Compiles.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/Compiles.cs:788]:
```suggestion
```

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:441]: Unsure if this is worth quibbling about, but `CharSet.Ansi` on non-Windows is Utf8 if I recall correctly. Based on certain reg keys on Windows it can also be Utf8. I don't know if this makes a difference in this case, but it is something to note.

**Review by AaronRobinsonMSFT — COMMENTED** [review]
LGTM, but I defer to @elinor-fung and @jkoritzinsky for final sign-off.

**Review by elinor-fung — APPROVED** [review]
The change looks good to me. I do want to make sure we understand the down-level support and dotnet/runtime-only downlevel support behaviour/expectations previously commented on though:

> > This is only applicable on the dotnet/runtime-only downlevel support, where we'll generate a source-generated stub for whatever parameters we can and forward whatever parameters we can't marshal (technically we also do this in main, but we emit errors in that case).
> 
> Is that true? It looks like we have tests that expect a forwarder stub if we can't marshal all the parameters:

I think it'd be good to have documentation and tests (or just doc for dotnet/runtime-only, since I don't think we have a way to test that) for the expected behaviour when not all parameters can be marshalled:
- when do we print a simple forwarder stub versus marshal what we can and forward the rest
- do we emit a diagnostic on what we can't marshal or just forward silently
- what is the difference in behaviour for this scenario in down-level and runtime-only down-level support (also non-down-level / current version, but I think we have a decent handle on and coverage of that)

The case of 'we generate a stub that marshals the parameters we can and forwards the remainder, without erroring' was surprising to me.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs
Comment (jtschuster) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportGenerator.cs:441]: It should error in down-level forwarding of UTF-8 and StringMarshalling.Custom, but this code path is taken in the normal path of LibraryImport in net7+, so we don't want to unconditionally warn here. Ideally this should warn when we hit this path for down-level support with partial marshalling, but we don't right now either.

**Review by jkoritzinsky — APPROVED** [review]
LGTM once @elinor-fung's comments are resolved.

**Comment by AaronRobinsonMSFT** [comment]
@jtschuster Are we waiting on something specific for this PR to go in?

**Comment by stephentoub** [comment]
Was this superceded by https://github.com/dotnet/runtime/pull/104416?

**Comment by AaronRobinsonMSFT** [comment]
> Was this superceded by #104416?

Yes. Closing.

**Pull request closed without merging**
